Multiple foreign keys with #[belongs_to], conflicting implementations of trait

I have a table `posts`, and every post has an author (`user_id`) and a possible parent post that it replies to:
```rust
#[derive(Queryable, Identifiable, Clone, Associations, Debug)]
#[belongs_to(User)]
#[belongs_to(Post, foreign_key="parent")]
pub struct Post {
	pub id: i64,
	pub user_id: i64,
	pub parent: Option<i64>,
	// ...
}
```
But I get:
> error[E0119]: conflicting implementations of trait `diesel::SelectableExpression<diesel::query_source::InnerJoinSource<models::schema::__diesel_infer_schema::infer_posts::posts::table, models::schema::__diesel_infer_schema::infer_posts::posts::table>, diesel::types::BigInt>` for type `models::schema::__diesel_infer_schema::infer_posts::posts::columns::id`:

What's the correct way to declare multiple foreign keys for this kind of usecase?

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -18,6 +18,11 @@ for Rust libraries in [RFC #1105](https://github.com/rust-lang/rfcs/blob/master/
   array that contains null, but a `table!` definition which specifies a type of
   `Array<Nullable<X>>` can now be deserialized to `Vec<Option<T>>`
 
+* [`#[belongs_to]`][belongs-to-0.11.0] associations can now be self referential.
+  This will generate the code required for
+  [`belonging_to`][belonging-to-0.11.0], without generating code for performing
+  a join.
+
 ### Changed
 
 * It is no longer possible to exhaustively match against

--- a/diesel/src/macros/associations/belongs_to.rs
+++ b/diesel/src/macros/associations/belongs_to.rs
@@ -226,25 +226,27 @@ macro_rules! BelongsTo {
             $($rest:tt)*
         })+],
     ) => {
-        joinable_inner!(
-            left_table_ty = $child_table_name::table,
-            right_table_ty = <$parent_struct as $crate::associations::HasTable>::Table,
-            right_table_expr = <$parent_struct as $crate::associations::HasTable>::table(),
-            foreign_key = $child_table_name::$foreign_key_name,
-            primary_key_ty = <<$parent_struct as $crate::associations::HasTable>::Table as $crate::Table>::PrimaryKey,
-            primary_key_expr = $crate::Table::primary_key(&<$parent_struct as $crate::associations::HasTable>::table()),
-        );
+        static_cond!(if $struct_name != $parent_struct {
+            joinable_inner!(
+                left_table_ty = $child_table_name::table,
+                right_table_ty = <$parent_struct as $crate::associations::HasTable>::Table,
+                right_table_expr = <$parent_struct as $crate::associations::HasTable>::table(),
+                foreign_key = $child_table_name::$foreign_key_name,
+                primary_key_ty = <<$parent_struct as $crate::associations::HasTable>::Table as $crate::Table>::PrimaryKey,
+                primary_key_expr = $crate::Table::primary_key(&<$parent_struct as $crate::associations::HasTable>::table()),
+            );
 
-        $(select_column_inner!(
-            $child_table_name::table,
-            <$parent_struct as $crate::associations::HasTable>::Table,
-            $child_table_name::$column_name,
-        );)+
-        select_column_inner!(
-            $child_table_name::table,
-            <$parent_struct as $crate::associations::HasTable>::Table,
-            $child_table_name::star,
-        );
+            $(select_column_inner!(
+                $child_table_name::table,
+                <$parent_struct as $crate::associations::HasTable>::Table,
+                $child_table_name::$column_name,
+            );)+
+            select_column_inner!(
+                $child_table_name::table,
+                <$parent_struct as $crate::associations::HasTable>::Table,
+                $child_table_name::star,
+            );
+        });
     };
 
     // Handle struct with no generics

--- a/diesel_tests/tests/associations.rs
+++ b/diesel_tests/tests/associations.rs
@@ -135,6 +135,42 @@ fn associations_can_be_grouped_multiple_levels_deep() {
     assert_eq!(expected_data, data);
 }
 
+#[test]
+fn self_referencing_associations() {
+    #[derive(Insertable, Queryable, Associations, Identifiable)]
+    #[derive(Debug, Clone, Copy, PartialEq)]
+    #[table_name="trees"]
+    #[belongs_to(Tree, foreign_key="parent_id")]
+    struct Tree {
+        id: i32,
+        parent_id: Option<i32>,
+    }
+
+    let conn = connection();
+    let test_data = vec![
+        Tree { id: 1, parent_id: None },
+        Tree { id: 2, parent_id: None },
+        Tree { id: 3, parent_id: Some(1) },
+        Tree { id: 4, parent_id: Some(2) },
+        Tree { id: 5, parent_id: Some(1) },
+    ];
+    insert(&test_data).into(trees::table)
+        .execute(&conn).unwrap();
+
+    let parents = trees::table.filter(trees::parent_id.is_null())
+        .load::<Tree>(&conn).unwrap();
+    let children = Tree::belonging_to(&parents)
+        .load::<Tree>(&conn).unwrap();
+    let children = children.grouped_by(&parents);
+    let data = parents.into_iter().zip(children).collect::<Vec<_>>();
+
+    let expected_data = vec![
+        (test_data[0], vec![test_data[2], test_data[4]]),
+        (test_data[1], vec![test_data[3]]),
+    ];
+    assert_eq!(expected_data, data);
+}
+
 fn conn_with_test_data() -> (TestConnection, User, User, User) {
     let connection = connection_with_sean_and_tess_in_users_table();
     insert(&NewUser::new("Jim", None)).into(users::table).execute(&connection).unwrap();

--- a/migrations/mysql/20170215170122_create_trees/down.sql
+++ b/migrations/mysql/20170215170122_create_trees/down.sql
@@ -1,0 +1,1 @@
+DROP TABLE trees;

--- a/migrations/mysql/20170215170122_create_trees/up.sql
+++ b/migrations/mysql/20170215170122_create_trees/up.sql
@@ -1,0 +1,4 @@
+CREATE TABLE trees (
+  id INTEGER PRIMARY KEY,
+  parent_id INTEGER
+);

--- a/migrations/postgresql/20170215170122_create_trees/down.sql
+++ b/migrations/postgresql/20170215170122_create_trees/down.sql
@@ -1,0 +1,1 @@
+DROP TABLE trees;

--- a/migrations/postgresql/20170215170122_create_trees/up.sql
+++ b/migrations/postgresql/20170215170122_create_trees/up.sql
@@ -1,0 +1,4 @@
+CREATE TABLE trees (
+  id SERIAL PRIMARY KEY,
+  parent_id INTEGER
+);

--- a/migrations/sqlite/20170215170122_create_trees/down.sql
+++ b/migrations/sqlite/20170215170122_create_trees/down.sql
@@ -1,0 +1,1 @@
+DROP TABLE trees;

--- a/migrations/sqlite/20170215170122_create_trees/up.sql
+++ b/migrations/sqlite/20170215170122_create_trees/up.sql
@@ -1,0 +1,4 @@
+CREATE TABLE trees (
+  id INTEGER PRIMARY KEY NOT NULL,
+  parent_id INTEGER
+);

